File: ios/Classes/PlayerView.m
//
//  PlayerView.m
//  flutter_aliplayer
//
//  Created by Mao Yanfei on 2020/8/31.
//

#import "PlayerView.h"
#import <AliyunPlayer/AliPlayer.h>

@interface PlayerView()<AVPDelegate>

@property(nonatomic)FlutterMethodChannel* methodChannel;
@property(nonatomic)AliPlayer * aliPlayer;
@property(nonatomic)UIView * renderView;

@end

@implementation PlayerView

- (instancetype)initWithWithFrame:(CGRect)frame viewIdentifier:(int64_t)viewId arguments:(id)args binaryMessenger:(NSObject<FlutterBinaryMessenger> *)messenger {
    if ([super init]) {
        self.renderView = [[UIView alloc] initWithFrame:frame];
        self.methodChannel = [FlutterMethodChannel
                              methodChannelWithName:[@"li.haomin.aliplayer.view/player" stringByAppendingFormat:@"_%lld", viewId]
              binaryMessenger:messenger];
        __block typeof(self) weakSelf = self;
        [self.methodChannel setMethodCallHandler:^(FlutterMethodCall *call, FlutterResult result) {
            [weakSelf handleMethodCall:call result:result];
        }];
    }
    return self;
}

- (UIView *)view {
    return self.renderView;
}

- (void) handleMethodCall:(FlutterMethodCall *) call result:(FlutterResult) result {
    
    NSDictionary *dic = (NSDictionary *) call.arguments;
    
    if ([@"init" isEqualToString:call.method]) {
        if (_aliPlayer != nil) {
            return;
        }
        _aliPlayer = [[AliPlayer alloc] init];
        _aliPlayer.playerView = _renderView;
        _aliPlayer.delegate = self;
        _aliPlayer.autoPlay = YES;
    } else if ([@"setDataSource" isEqualToString:call.method]) {
        AVPVidAuthSource *source = [[AVPVidAuthSource alloc] init];
        source.vid = dic[@"vid"];
        source.playAuth = dic[@"auth"];
        [_aliPlayer setAuthSource:source];
        [_aliPlayer prepare];
    } else if ([@"start" isEqualToString:call.method]) {
        [_aliPlayer start];
    } else if ([@"stop" isEqualToString:call.method]) {
        [_aliPlayer stop];
    } else if ([@"pause" isEqualToString:call.method]) {
        [_aliPlayer pause];
    } else if ([@"seekTo" isEqualToString:call.method]) {
        [_aliPlayer seekToTime:[dic[@"seek"] longLongValue] seekMode:AVP_SEEKMODE_INACCURATE];
    } else if ([@"release" isEqualToString:call.method]) {
        if (_aliPlayer != nil) {
            [_aliPlayer destroy];
            _aliPlayer = nil;
        }
    } else if ([@"getDuration" isEqualToString:call.method]) {
        result(@(_aliPlayer.duration));
    } else if ([@"getVideoWidth" isEqualToString:call.method]) {
        result(@(_aliPlayer.width));
    } else if ([@"getVideoHeight" isEqualToString:call.method]) {
        result(@(_aliPlayer.height));
    } else if ([@"setVolume" isEqualToString:call.method]) {
        _aliPlayer.volume = [dic[@"volume"] floatValue];
    } else if ([@"getVolume" isEqualToString:call.method]) {
        result(@(_aliPlayer.volume));
    } else if ([@"setMute" isEqualToString:call.method]) {
        _aliPlayer.muted = [dic[@"flag"] boolValue];
    } else if ([@"isMute" isEqualToString:call.method]) {
        result(@(_aliPlayer.isMuted));
    } else if ([@"redraw" isEqualToString:call.method]) {
        [_aliPlayer redraw];
    } else if ([@"reset" isEqualToString:call.method]) {
           [_aliPlayer reset];
    } else if ([@"reload" isEqualToString:call.method]) {
           [_aliPlayer reload];
    } else {
      result(FlutterMethodNotImplemented);
    }
}

- (void)onPlayerEvent:(AliPlayer *)player eventType:(AVPEventType)eventType {
    switch (eventType) {
        case AVPEventPrepareDone:
            [_methodChannel invokeMethod:@"eventListener" arguments:@{
                @"event": @"onPrepared"
            }];
            break;
        case AVPEventCompletion:
            [_methodChannel invokeMethod:@"eventListener" arguments:@{
                @"event": @"onCompletion"
            }];
            break;
        case AVPEventLoadingStart:
            [_methodChannel invokeMethod:@"eventListener" arguments:@{
                @"event": @"onLoadingBegin"
            }];
            break;
        case AVPEventLoadingEnd:
            [_methodChannel invokeMethod:@"eventListener" arguments:@{
                @"event": @"onLoadingEnd"
            }];
            break;
            
        default:
            break;
    }
}

- (void)onPlayerStatusChanged:(AliPlayer *)player oldStatus:(AVPStatus)oldStatus newStatus:(AVPStatus)newStatus {
    [_methodChannel invokeMethod:@"eventListener" arguments:@{
        @"event": @"onStateChanged",
        @"state": @(newStatus)
    }];
}

- (void)onBufferedPositionUpdate:(AliPlayer *)player position:(int64_t)position {
    [_methodChannel invokeMethod:@"eventListener" arguments:@{
        @"event": @"onBufferedPosition",
        @"position": @(position)
    }];
}

- (void)onCurrentPositionUpdate:(AliPlayer *)player position:(int64_t)position {
    [_methodChannel invokeMethod:@"eventListener" arguments:@{
        @"event": @"onCurrentPosition",
        @"position": @(position)
    }];
}

- (void)onLoadingProgress:(AliPlayer *)player progress:(float)progress {
    NSLog(@"onLoadingProgress");
    [_methodChannel invokeMethod:@"eventListener" arguments:@{
        @"event": @"onLoadingProgress",
        @"percent": @(progress),
        @"kbps": @(0)
    }];
}

- (void)onError:(AliPlayer *)player errorModel:(AVPErrorModel *)errorModel {
    [_methodChannel invokeMethod:@"eventListener" arguments:@{
        @"event": @"onError",
        @"code": @(errorModel.code),
        @"msg": errorModel.message
    }];
}

@end

File: android/src/main/java/li/haomin/flutter_aliplayer/view/PlayerView.java
package li.haomin.flutter_aliplayer.view;

import android.content.Context;
import android.view.SurfaceView;
import android.view.View;

import com.aliyun.player.AliPlayer;
import com.aliyun.player.AliPlayerFactory;
import com.aliyun.player.source.VidAuth;
import com.aliyun.player.source.UrlSource;

import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.StandardMessageCodec;
import io.flutter.plugin.platform.PlatformView;
import io.flutter.plugin.platform.PlatformViewFactory;
import li.haomin.flutter_aliplayer.listener.EventListener;

public class PlayerView extends PlatformViewFactory implements PlatformView, MethodChannel.MethodCallHandler {
    public static final String VIEW_NAME = "li.haomin.aliplayer.view/player";

    private Context context;
    private BinaryMessenger messenger;

    private MethodChannel methodChannel;
    private EventListener eventListener;

    private SurfaceView surfaceView;
    private AliPlayer aliPlayer;

    public PlayerView(Context context, BinaryMessenger messenger) {
        super(StandardMessageCodec.INSTANCE);
        this.context = context;
        this.messenger = messenger;
    }

    private PlayerView(Context context) {
        super(StandardMessageCodec.INSTANCE);
        this.context = context;
        surfaceView = new SurfaceView(this.context);
    }

    @Override
    public View getView() {
        return surfaceView;
    }

    @Override
    public void dispose() {

    }

    @Override
    public PlatformView create(Context context, int viewId, Object args) {
        Map<String, Object> params = (Map<String, Object>) args;
        PlayerView playerView = new PlayerView(context);
        // 绑定方法监听器
        methodChannel = new MethodChannel(messenger, VIEW_NAME + "_" + viewId);
        methodChannel.setMethodCallHandler(playerView);
        playerView.setListener(new EventListener(methodChannel));
        return playerView;
    }

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        switch (call.method) {
            case "init":
                init();
                break;
            case "setDataSource":
                VidAuth vidAuth = new VidAuth();
                vidAuth.setVid(call.argument("vid").toString());
                vidAuth.setPlayAuth(call.argument("auth").toString());
                aliPlayer.setDataSource(vidAuth);
                aliPlayer.setAutoPlay(true);
                aliPlayer.prepare();
                break;
            case "setUrlDataSource":
                UrlSource urlSource = new UrlSource();
                urlSource.setUri(call.argument("playurl").toString());
                aliPlayer.setDataSource(urlSource);
                aliPlayer.setAutoPlay(true);
                aliPlayer.prepare();
                break;
            case "start":
                aliPlayer.start();
                break;
            case "stop":
                aliPlayer.stop();
                break;
            case "pause":
                aliPlayer.pause();
                break;
            case "seekTo":
                long seek = Long.valueOf(call.argument("seek").toString());
                aliPlayer.seekTo(seek);
                break;
            case "release":
                if (aliPlayer != null) {
                    aliPlayer.release();
                    aliPlayer = null;
                }
                break;
            case "getDuration":
                result.success(aliPlayer.getDuration());
                break;
            case "getVideoWidth":
                result.success(aliPlayer.getVideoWidth());
                break;
            case "getVideoHeight":
                result.success(aliPlayer.getVideoHeight());
                break;
            case "setVolume":
                float volume = Float.valueOf(call.argument("volume").toString());
                aliPlayer.setVolume(volume);
                break;
            case "getVolume":
                result.success(aliPlayer.getVolume());
                break;
            case "setMute":
                boolean flag = call.argument("flag");
                aliPlayer.setMute(flag);
                break;
            case "isMute":
                result.success(aliPlayer.isMute());
                break;
            default:
                result.notImplemented();
        }
    }

    private void init() {
        if (aliPlayer != null) {
            return;
        }
        aliPlayer = AliPlayerFactory.createAliPlayer(context);
        aliPlayer.setOnPreparedListener(eventListener);
        aliPlayer.setOnCompletionListener(eventListener);
        aliPlayer.setOnErrorListener(eventListener);
        aliPlayer.setOnStateChangedListener(eventListener);
        aliPlayer.setOnInfoListener(eventListener);
        aliPlayer.setOnLoadingStatusListener(eventListener);
        aliPlayer.setDisplay(surfaceView.getHolder());
        aliPlayer.setAutoPlay(true);
    }

    private void setListener(EventListener eventListener) {
        this.eventListener = eventListener;
    }
}
